File: core/commands/registry/irules/datagram__tcp.py
# Enriched from F5 iRules reference documentation.
"""DATAGRAM::tcp -- Returns TCP header and payload information."""

from __future__ import annotations

from ....compiler.side_effects import ConnectionSide, SideEffect, SideEffectTarget
from .._base import CommandDef, make_av
from ..models import CommandSpec, FormKind, FormSpec, HoverSnippet, ValidationSpec
from ..namespace_models import EventRequires
from ..signatures import Arity
from ._base import _IRULES_ONLY, register

_SOURCE = "https://clouddocs.f5.com/api/irules/DATAGRAM__tcp.html"


_av = make_av(_SOURCE)


@register
class DatagramTcpCommand(CommandDef):
    name = "DATAGRAM::tcp"

    @classmethod
    def spec(cls) -> CommandSpec:
        return CommandSpec(
            name="DATAGRAM::tcp",
            dialects=_IRULES_ONLY,
            hover=HoverSnippet(
                summary="Returns TCP header and payload information.",
                synopsis=(
                    "DATAGRAM::tcp (flags | payload_length | window)",
                    "DATAGRAM::tcp (option | option_count) (IPV4_OPTION)?",
                    "DATAGRAM::tcp payload (LENGTH)?",
                ),
                snippet=(
                    "This iRules command returns tcp header information.\n"
                    "Note: throws an error if L4 protocol of the current connection is not\n"
                    "TCP\n"
                    "\n"
                    "DATAGRAM::tcp flags\n"
                    "\n"
                    "     * This command returns TCP header flags as an integer value.\n"
                    "\n"
                    "DATAGRAM::tcp option\n"
                    "\n"
                    "     * This command returns a Tcl list of TCP options. See\n"
                    "       DATAGRAM::ip option for details on behavior.\n"
                    "\n"
                    "DATAGRAM::tcp option [option-code]\n"
                    "\n"
                    "     * This command returns a Tcl list of TCP option values for TCP option\n"
                    "       with a given option code. See DATAGRAM::ip option for details\n"
                    "       on behavior."
                ),
                source=_SOURCE,
                examples=(
                    "when FLOW_INIT {\n"
                    "  if { [IP::protocol] == 6 } {\n"
                    '    log local0. "TCP Flow: [IP::client_addr] [TCP::client_port] --> [IP::local_addr] [TCP::local_port]"\n'
                    '    log local0. "TCP Payload Length = [DATAGRAM::tcp payload_length] Payload: [DATAGRAM::tcp payload 100]"\n'
                    "  }\n"
                    "}"
                ),
            ),
            forms=(
                FormSpec(
                    kind=FormKind.DEFAULT,
                    synopsis="DATAGRAM::tcp (flags | payload_length | window)",
                    arg_values={
                        0: (
                            _av(
                                "flags",
                                "DATAGRAM::tcp flags",
                                "DATAGRAM::tcp (flags | payload_length | window)",
                            ),
                            _av(
                                "payload_length",
                                "DATAGRAM::tcp payload_length",
                                "DATAGRAM::tcp (flags | payload_length | window)",
                            ),
                            _av(
                                "window",
                                "DATAGRAM::tcp window",
                                "DATAGRAM::tcp (flags | payload_length | window)",
                            ),
                            _av(
                                "option",
                                "DATAGRAM::tcp option",
                                "DATAGRAM::tcp (option | option_count) (IPV4_OPTION)?",
                            ),
                            _av(
                                "option_count",
                                "DATAGRAM::tcp option_count",
                                "DATAGRAM::tcp (option | option_count) (IPV4_OPTION)?",
                            ),
                        )
                    },
                ),
            ),
            validation=ValidationSpec(
                arity=Arity(),
            ),
            event_requires=EventRequires(
                profiles=frozenset({"FLOW"}), also_in=frozenset({"CLIENT_DATA"})
            ),
            side_effect_hints=(
                SideEffect(
                    target=SideEffectTarget.NETWORK_IO,
                    reads=True,
                    connection_side=ConnectionSide.BOTH,
                ),
            ),
        )

File: rust/tcl-registry/src/commands/irules/datagram__tcp.rs
// tcl-lsp — a language server and toolchain for Tcl
// Copyright (C) 2026 James Deucker (bitwisecook) <https://github.com/bitwisecook>
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//
// SPDX-License-Identifier: AGPL-3.0-or-later

//! `DATAGRAM::tcp` iRules command.
use crate::prelude::*;
pub const fn spec() -> CommandSpec {
    CommandSpec {
        name: "DATAGRAM::tcp",
        dialects: Some(DialectSet::IRULES),
        arity: Arity::at_least(0),
        hover: Some(HoverSnippet {
            summary: "Returns TCP header and payload information.",
            synopsis: &[
                "DATAGRAM::tcp (flags | payload_length | window)",
                "DATAGRAM::tcp (option | option_count) (IPV4_OPTION)?",
                "DATAGRAM::tcp payload (LENGTH)?",
            ],
            snippet: "This iRules command returns tcp header information.\nNote: throws an error if L4 protocol of the current connection is not\nTCP\n\nDATAGRAM::tcp flags\n\n     * This command returns TCP header flags as an integer value.\n\nDATAGRAM::tcp option\n\n     * This command returns a Tcl list of TCP options. See\n       DATAGRAM::ip option for details on behavior.\n\nDATAGRAM::tcp option [option-code]\n\n     * This command returns a Tcl list of TCP option values for TCP option\n       with a given option code. See DATAGRAM::ip option for details\n       on behavior.",
            source: "https://clouddocs.f5.com/api/irules/DATAGRAM__tcp.html",
            examples: "when FLOW_INIT {\n  if { [IP::protocol] == 6 } {\n    log local0. \"TCP Flow: [IP::client_addr] [TCP::client_port] --> [IP::local_addr] [TCP::local_port]\"\n    log local0. \"TCP Payload Length = [DATAGRAM::tcp payload_length] Payload: [DATAGRAM::tcp payload 100]\"\n  }\n}",
            return_value: "",
        }),
        event_requires: Some(EventRequires {
            client_side: false,
            server_side: false,
            transport: None,
            profiles: &["FLOW"],
            also_in: &["CLIENT_DATA"],
            init_only: false,
            flow: false,
            capability: None,
        }),
        forms: &[FormSpec {
            synopsis: "DATAGRAM::tcp (flags | payload_length | window)",
            ..FormSpec::DEFAULT
        }],
        side_effects: &[SideEffect {
            target: SideEffectTarget::NetworkIo,
            reads: true,
            connection_side: ConnectionSide::Both,
            ..SideEffect::DEFAULT
        }],
        ..CommandSpec::DEFAULT
    }
}
